### ISSUE TITLE
Adblocker affects quality for non ad streams in some cases

### ISSUE THREAD

**Issue by towlie** [saucettv/VideoAdBlockForTwitch#32]
Version: 5.3.1
How to reproduce:
Open a stream with a preroll ad playing and a second stream where no ads will be played.

The second stream will be set down to 480p even when there is no ad. And will not set it to the source (higher) quality until you refresh the stream.

This can also lead to a new problem, if you close the ad stream before the ad blocker is done blocking. (locked 480p quality until the user changes it manually)

**Comment by saucettv** [comment]
Thank you for the report. That's strange, I thought the content script would work individually per tab. I will take a look and see if anything can be done. I have noticed you can manually up the quality on the stream that shouldn't have an ad, but yeah, it is annoying.

**Comment by saucettv** [comment]
Actually thinking about it, this is how it will have to work for now. The script has to force the player quality to 480p during any ad, otherwise the stream will freeze on the tab where the ad shows. Twitch then uses the quality that is set on the player for the new tab, so until the script reverts the quality, it will be 480p on the other tab, unless manually changed. I will having a think about it, but this might be just an issue we will have to put up with.

**Comment by pixeltris** [comment]
I haven't been able to look at that original freezing issue, but how about doing something like this on each m3u8 change:

```
var currentQualityIndex = -1;
var qualities = player.getQualities();
var currentQuality = player.getQuality();
if (qualities && currentQuality && currentQuality.name) {
    for (var i = 0; i < qualities.length; i++) {
        if (qualities[i].name == currentQuality.name) {
            currentQualityIndex = i;
            break;
        }
    }
}
if (currentQualityIndex >= 0) {
    player.setQuality(player.getQualities()[currentQualityIndex]);
}
```

Or surely just a pause/play would also work on each m3u8 change. Albeit this is kind of annoying as that would be 3 minor interruptions:

- When entering the ad
- When finding an ad-free stream while there are still ad segments in the original stream
- Back to the original steam

Might be able to get that down to 2 by either extending the low quality duration or permanently swapping out the urls with the ones found mid-ad.

**Comment by saucettv** [comment]
I've tried lot's of things now. The setQuality method doesn't work for the freezing streams. Pause/play will only work if we swap out the entire m3u8 on the player somehow. I just don't have enough time to work on it that much.

The issue at the moment isn't really that bad. It only happens if you open lots of tabs all at the same time and one of first tabs gets an ad, causing the player to set itself to 480p, so then the new tabs will be 480p, because the ads are still being blocked. Users can either wait for the original tab to finish blocking or just manually change the quality on the new tabs, if opening them that fast.

**Comment by pixeltris** [comment]
Yea I couldn't get anything to work nicely. I created a solution which swaps out the main m3u8 when the stream transition between ads, but is otherwise unmodified. Haven't done a huge amount of testing but it seems to work (though the player reload is generally a bit slower than a pause/play or quality change). The code / some additional details can be found here https://github.com/pixeltris/TwitchAdSolutions/issues/59

**Comment by saucettv** [comment]
That's cool, thanks. I will prob try and make things work a little better in the future but for now, the new tab being 480p if an ad is already running on another tab isn't that big of a problem for me, so think I will leave it for now. Worst case is that the user has to manually change the quality on new tabs that are opened while an ad plays.

**Comment by pixeltris** [comment]
Another way you might be able to reduce the impact of the quality change is by doing something like this:

```
realSetItem = localStorage.setItem;
localStorage.setItem = function(keyName, keyValue) {
    if (keyName == 'video-quality' && isTemporaryQualityChange) {
        return;
    }
    realSetItem.apply(this, arguments);
};
```

Where you'd set `isTemporaryQualityChange` to true while you're doing the quality changes via code. This should prevent other tabs / new streams from going to 480p.

**Comment by saucettv** [comment]
I like that, thanks pixeltris, I will try implement it when I have time :)

**Comment by saucettv** [comment]
Added the fix to 5.3.3.
